### PR TITLE
Refactor WMRUtilities to support Unity legacy XR and XR SDK

### PR DESCRIPTION
**Pull request by keveleigh**
## Overview
Refactors the static `WindowsMixedRealityUtilities` to take in a provider, which can be set in platform-specific data providers. This allows the utilities class to be called generically regardless of the Unity XR backend.

## Changes
- Fixes: the current build break when using DotNetWinRT.

### CODE REVIEW

File: Assets/MixedRealityToolkit.Providers/WindowsMixedReality/Shared/Definitions/IWindowsMixedRealityUtilitiesProvider.cs [hunk 1]
```diff
@@ -0,0 +1,20 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+
+namespace Microsoft.MixedReality.Toolkit.WindowsMixedReality
+{
+    public interface IWindowsMixedRealityUtilitiesProvider
```
Comment (wiwei) [inline Assets/MixedRealityToolkit.Providers/WindowsMixedReality/Shared/Definitions/IWindowsMixedRealityUtilitiesProvider.cs]: Nit: Might also be worth adding some doccomments to this to (if only to help future generations understand what should go on this interface, and what shouldn't)

It's pretty clear from the contents what the things inside of it are now, though these are the types of interfaces that can balloon to much larger things going forward without guidance about initial intent (i.e. maybe ballooning is also fine, etc)

File: Assets/MixedRealityToolkit.Providers/WindowsMixedReality/Shared/WindowsMixedRealityUtilities.cs [hunk 2]
```diff
@@ -18,7 +17,7 @@
 #endif
 #endif // (UNITY_WSA && DOTNETWINRT_PRESENT) || WINDOWS_UWP
 
-namespace Microsoft.MixedReality.Toolkit.WindowsMixedReality.Input
+namespace Microsoft.MixedReality.Toolkit.WindowsMixedReality
```
Comment (wiwei) [inline Assets/MixedRealityToolkit.Providers/WindowsMixedReality/Shared/WindowsMixedRealityUtilities.cs]: Adding a note here with a link to https://github.com/microsoft/MixedRealityToolkit-Unity/issues/6922 so that we capture this namespace change on this thing.

File: Assets/MixedRealityToolkit.Providers/WindowsMixedReality/WindowsMixedRealityUtilitiesProvider.cs [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+using System;
+using System.Runtime.InteropServices;
+
+#if UNITY_WSA
+using UnityEngine.XR.WSA;
+#endif // UNITY_WSA
+
+namespace Microsoft.MixedReality.Toolkit.WindowsMixedReality
+{
+    public class WindowsMixedRealityUtilitiesProvider : IWindowsMixedRealityUtilitiesProvider
```
Comment (wiwei) [inline Assets/MixedRealityToolkit.Providers/WindowsMixedReality/WindowsMixedRealityUtilitiesProvider.cs]: Is the XR SDK version of this going to be named something else (i.e. XrSdkMixedRealityUtilitiesProvider)? If so, I wonder if the interface itself IWindowsMixedRealityUtilitiesProvider should be named something a bit more generic (i.e. IMixedRealityReferenceFrameProvider). As it's named it feels like there's mostly just this 1-1 mapping of IWindowsMixedRealityUtilitiesProvider ->WindowsMixedRealityUtilitiesProvider, but maybe seeing the other implementation of this would make it much clearer.

File: Assets/MixedRealityToolkit.Providers/WindowsMixedReality/Shared/WindowsMixedRealityUtilities.cs (hunk 2 above)
Comment (wiwei) [reply Assets/MixedRealityToolkit.Providers/WindowsMixedReality/Shared/WindowsMixedRealityUtilities.cs]: Mostly that we should have a write up and notes about this change, because it is technically breaking (I know that other consumers that we've done work for this stuff may end up using this directly, so it's something we should call out).

@davidkline-ms  thoughts?

File: Assets/MixedRealityToolkit.Providers/WindowsMixedReality/Shared/Definitions/IWindowsMixedRealityUtilitiesProvider.cs (hunk 1 above)
Comment (keveleigh) [reply Assets/MixedRealityToolkit.Providers/WindowsMixedReality/Shared/Definitions/IWindowsMixedRealityUtilitiesProvider.cs]: This one shouldn't increase very much, if at all, and really only needs to exist while we support both legacy Unity XR and XR SDK.

I'll add some comments though!

File: Assets/MixedRealityToolkit.Providers/WindowsMixedReality/Shared/WindowsMixedRealityUtilities.cs (hunk 2 above)
Comment (keveleigh) [reply Assets/MixedRealityToolkit.Providers/WindowsMixedReality/Shared/WindowsMixedRealityUtilities.cs]: Yeah, I waffled on actually changing this namespace or not. It makes it more "correct", but is a breaking change, like you pointed out. I can revert

File: Assets/MixedRealityToolkit.Providers/WindowsMixedReality/WindowsMixedRealityUtilitiesProvider.cs (hunk 3 above)
Comment (keveleigh) [reply Assets/MixedRealityToolkit.Providers/WindowsMixedReality/WindowsMixedRealityUtilitiesProvider.cs]: Hmm...good point. I left the full `WindowsMixedReality` in there since it's WMR-specific, which would probably make the other provider something long like...`WindowsMixedRealityXRSDKUtilitiesProvider` or `XRSDKWindowsMixedRealityUtilitiesProvider`, which matches my naming so far but is a bit wordy.

I've been keeping scripts which use the existing XR pipeline with the same naming patterns we've been using (as opposed to adding `Legacy` or something) and adding `XRSDK` to differentiate.

File: Assets/MixedRealityToolkit.Providers/WindowsMixedReality/Shared/Definitions/IWindowsMixedRealityUtilitiesProvider.cs (hunk 1 above)
Comment (keveleigh) [reply Assets/MixedRealityToolkit.Providers/WindowsMixedReality/Shared/Definitions/IWindowsMixedRealityUtilitiesProvider.cs]: Added a summary with some remarks

File: Assets/MixedRealityToolkit.Providers/WindowsMixedReality/Definitions/HolographicFrameNativeData.cs [hunk 4]
```diff
@@ -24,12 +24,12 @@ public struct HolographicFrameNativeData
         public uint MaxNumberOfCameras;
 
         /// <summary>
-        /// The current native root <see href="https://docs.microsoft.com/uwp/api/windows.perception.spatial.spatialcoordinatesystem">ISpatialCoordinateSystem</see>).
+        /// The current native root <see href="https://docs.microsoft.com/uwp/api/windows.perception.spatial.spatialcoordinatesystem">ISpatialCoordinateSystem</see>.
```
Comment (davidkline-ms) [inline Assets/MixedRealityToolkit.Providers/WindowsMixedReality/Definitions/HolographicFrameNativeData.cs]: is this the root pointer for all native access? not sure what "root" means here

File: Assets/MixedRealityToolkit.Providers/WindowsMixedReality/Definitions/HolographicFrameNativeData.cs [hunk 5]
```diff
@@ -24,12 +24,12 @@ public struct HolographicFrameNativeData
...
         /// </summary>
         public IntPtr ISpatialCoordinateSystemPtr;
         
         /// <summary>
-        /// The current native <see href="https://docs.microsoft.com/uwp/api/Windows.Graphics.Holographic.HolographicFrame">IHolographicFrame</see>).
+        /// The current native <see href="https://docs.microsoft.com/uwp/api/Windows.Graphics.Holographic.HolographicFrame">IHolographicFrame</see>.
```
Comment (davidkline-ms) [inline Assets/MixedRealityToolkit.Providers/WindowsMixedReality/Definitions/HolographicFrameNativeData.cs]:
```suggestion
        /// The current native frame <see href="https://docs.microsoft.com/uwp/api/Windows.Graphics.Holographic.HolographicFrame">IHolographicFrame</see>.
```

File: Assets/MixedRealityToolkit.Providers/WindowsMixedReality/Shared/Definitions/IWindowsMixedRealityUtilitiesProvider.cs (hunk 1 above)
Comment (davidkline-ms) [reply Assets/MixedRealityToolkit.Providers/WindowsMixedReality/Shared/Definitions/IWindowsMixedRealityUtilitiesProvider.cs]: > really only needs to exist while we support both legacy Unity XR and XR SDK

That said, this is a good pattern to maintain going forward as the engine may change in ways that the provider needs to accomodate.

File: Assets/MixedRealityToolkit.Providers/WindowsMixedReality/Definitions/HolographicFrameNativeData.cs (hunk 4 above)
Comment (keveleigh) [reply Assets/MixedRealityToolkit.Providers/WindowsMixedReality/Definitions/HolographicFrameNativeData.cs]: It's the `ISpatialCoordinateSystem` at the root of the scene. Happy to use a different word if something else makes it more clear!

File: Assets/MixedRealityToolkit.Providers/WindowsMixedReality/Shared/Definitions/IWindowsMixedRealityUtilitiesProvider.cs [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    /// to provide access to specific underlying native objects relevant to Windows Mixed Reality.
+    /// </summary>
+    /// <remarks>
+    /// This is intended to be used to support both XR SDK and Unity's legacy XR pipeline, which provide
+    /// different APIs to access these native objects.
+    /// </remarks>
+    public interface IWindowsMixedRealityUtilitiesProvider
+    {
+        /// <summary>
+        /// The current native root <see href="https://docs.microsoft.com/uwp/api/windows.perception.spatial.spatialcoordinatesystem">ISpatialCoordinateSystem</see>.
```
Comment (davidkline-ms) [inline Assets/MixedRealityToolkit.Providers/WindowsMixedReality/Shared/Definitions/IWindowsMixedRealityUtilitiesProvider.cs]: please see above comments related to this and the next summary block

File: Assets/MixedRealityToolkit.Providers/WindowsMixedReality/Shared/WindowsMixedRealityUtilities.cs (hunk 2 above)
Comment (davidkline-ms) [reply Assets/MixedRealityToolkit.Providers/WindowsMixedReality/Shared/WindowsMixedRealityUtilities.cs]: i have been trying, in recent PRs to document changes as they happen. even so, adding a comment to #6922 will help ensure that the change notes are as complete as possible

File: Assets/MixedRealityToolkit.Providers/WindowsMixedReality/Shared/WindowsMixedRealityUtilities.cs [hunk 7]
```diff
@@ -51,6 +50,8 @@ private static SpatialCoordinateSystem GetSpatialCoordinateSystem(IntPtr nativeP
         }
 #endif //ENABLE_DOTNET
 
+        public static IWindowsMixedRealityUtilitiesProvider WmrUtilitiesProvider { get; set; } = null;
```
Comment (davidkline-ms) [inline Assets/MixedRealityToolkit.Providers/WindowsMixedReality/Shared/WindowsMixedRealityUtilities.cs]: Would there be any other utilities provider other than Windows Mixed Reality? If not, maybe name this as UtilitiesProvider for ease of reading?

File: Assets/MixedRealityToolkit.Providers/WindowsMixedReality/Definitions/HolographicFrameNativeData.cs (hunk 5 above)
Comment (keveleigh) [reply Assets/MixedRealityToolkit.Providers/WindowsMixedReality/Definitions/HolographicFrameNativeData.cs]: This change will resolve to `The current native frame IHolographicFrame` as opposed to just `The current native IHolographicFrame`, which I think is more clear.

File: Assets/MixedRealityToolkit.Providers/WindowsMixedReality/Shared/WindowsMixedRealityUtilities.cs (hunk 7 above)
Comment (keveleigh) [reply Assets/MixedRealityToolkit.Providers/WindowsMixedReality/Shared/WindowsMixedRealityUtilities.cs]: For this static class as it's currently named and written, no. This class is specific to Windows Mixed Reality. I can still rename it though, since it's a little redundant.

File: Assets/MixedRealityToolkit.Providers/WindowsMixedReality/Shared/WindowsMixedRealityUtilities.cs [hunk 8]
```diff
@@ -62,18 +63,17 @@ public static SpatialCoordinateSystem SpatialCoordinateSystem
         {
             get
             {
+                if (spatialCoordinateSystem == null && WmrUtilitiesProvider != null)
+                {
 #if ENABLE_DOTNET
-                return spatialCoordinateSystem ?? (spatialCoordinateSystem = GetSpatialCoordinateSystem(WorldManager.GetNativeISpatialCoordinateSystemPtr()));
+                    spatialCoordinateSystem = GetSpatialCoordinateSystem(WmrUtilitiesProvider.ISpatialCoordinateSystemPtr);
```
Comment (davidkline-ms) [inline Assets/MixedRealityToolkit.Providers/WindowsMixedReality/Shared/WindowsMixedRealityUtilities.cs]: Consider turning this into a function?

File: Assets/MixedRealityToolkit.Providers/WindowsMixedReality/Shared/WindowsMixedRealityUtilities.cs [hunk 9]
```diff
@@ -88,14 +88,15 @@ public static HolographicFrame CurrentHolographicFrame
         {
             get
             {
+                if (WmrUtilitiesProvider == null)
+                {
+                    return null;
+                }
+
 #if DOTNETWINRT_PRESENT
```
Comment (davidkline-ms) [inline Assets/MixedRealityToolkit.Providers/WindowsMixedReality/Shared/WindowsMixedRealityUtilities.cs]: no ENABLE_DOTNET here?

File: Assets/MixedRealityToolkit.Providers/WindowsMixedReality/WindowsMixedRealityUtilitiesProvider.cs (hunk 3 above)
Comment (davidkline-ms) [reply Assets/MixedRealityToolkit.Providers/WindowsMixedReality/WindowsMixedRealityUtilitiesProvider.cs]: Along the lines of @wiwei's question; should we have an overall XR SDK utilities provider that aggregates anything we need to access, regardless of platform?  Is that feasible?

File: Assets/MixedRealityToolkit.Providers/WindowsMixedReality/Shared/WindowsMixedRealityUtilities.cs (hunk 9 above)
Comment (keveleigh) [reply Assets/MixedRealityToolkit.Providers/WindowsMixedReality/Shared/WindowsMixedRealityUtilities.cs]: It may be needed here, but the existing .NET workaround was specific to marshaling the `ISpatialCoordinateSystemPtr`. It would need updating to support `IHolographicFramePtr`, and the DotNetWinRT adapter may provide a better workaround for .NET Native. Testing needed.

File: Assets/MixedRealityToolkit.Providers/WindowsMixedReality/Shared/WindowsMixedRealityUtilities.cs (hunk 7 above)
Comment (keveleigh) [reply Assets/MixedRealityToolkit.Providers/WindowsMixedReality/Shared/WindowsMixedRealityUtilities.cs]: Renamed to `UtilitiesProvider`

File: Assets/MixedRealityToolkit.Providers/WindowsMixedReality/Shared/WindowsMixedRealityUtilities.cs (hunk 2 above)
Comment (keveleigh) [reply Assets/MixedRealityToolkit.Providers/WindowsMixedReality/Shared/WindowsMixedRealityUtilities.cs]: Added a note to #6922

File: Assets/MixedRealityToolkit.Providers/WindowsMixedReality/Shared/WindowsMixedRealityUtilities.cs (hunk 8 above)
Comment (keveleigh) [reply Assets/MixedRealityToolkit.Providers/WindowsMixedReality/Shared/WindowsMixedRealityUtilities.cs]: I'm going to leave this as-is, since it's only called in one place and, although a bit confusing with all the `#ifdef`s, I'm not sure moving it into a function will help overall reading of the method.

File: Assets/MixedRealityToolkit.Providers/WindowsMixedReality/WindowsMixedRealityUtilitiesProvider.cs (hunk 3 above)
Comment (keveleigh) [reply Assets/MixedRealityToolkit.Providers/WindowsMixedReality/WindowsMixedRealityUtilitiesProvider.cs]: We may move to a pattern like that in the future, but I've been thinking of these `Utilities` as platform-specific extensions, where we're able to implement additional platform-specific features based on Unity's native pointers. I'm only familiar with the native pointers for WMR, though other platforms may have them. When we want/need to take advantage of another platform's native pointer in the future, we can best figure out how to do that then. In my opinion, making it explicitly platform-specific is my preferred pattern, since there may not be a straight forward abstraction for each concept.

File: Assets/MixedRealityToolkit.Providers/WindowsMixedReality/Definitions/HolographicFrameNativeData.cs (hunk 4 above)
Comment (davidkline-ms) [reply Assets/MixedRealityToolkit.Providers/WindowsMixedReality/Definitions/HolographicFrameNativeData.cs]: ignore me.... i missed the <see ... /> part of the comment. it is fine as is

Comment (LongForever1993) [reply Assets/MixedRealityToolkit.Providers/WindowsMixedReality/Definitions/HolographicFrameNativeData.cs]: Error:BaseWindowsMixedRealityCameraSettings.cs(54,17): error CS0246: The type or namespace name 'WindowsMixedRealityCameraSettingsProfile' could not be found (are you missing a using directive or an assembly reference?)

Comment (keveleigh) [reply Assets/MixedRealityToolkit.Providers/WindowsMixedReality/Definitions/HolographicFrameNativeData.cs]: @LongForever1993 Where and when are you seeing that error? Please file an issue if it's consistent.
